[체스 - 3단계] 냥인(최가희) 미션 제출합니다.

cutehumanS2 opened the pull request:
오리, 안녕하세요. 6기 냥인입니다.

제가 구현 속도가 느려서 이제야 3단계를 제출하게 되었네요.😿
이번 리뷰도 잘 부탁드립니다.

cutehumanS2 committed:
refactor: 기물별 이동 전략 교체

이동 방법에 따른 전략 -> 기물에 따른 전략으로 교체

cutehumanS2 committed:
fix: 찾는 기물이 존재하지 않는 경우 EMPTY 타입을 반환하도록 변경

cutehumanS2 committed:
docs(README): 추가 기능 목록 작성

cutehumanS2 committed:
feat: 점수 계산 기능

cutehumanS2 committed:
feat: 이긴 진영 산출 기능

+ 불필요한 메서드 제거

cutehumanS2 committed:
feat: 현재 이기고 있는 진영 산출 기능

cutehumanS2 committed:
feat: 게임 결과 출력 및 종료 기능

cutehumanS2 committed:
refactor: 보드 출력 로직 변경

- 2차원 배열 -> 리스트 활용
- 불필요한 메서드 제거
- 메서드명 및 파라미터명 변경
- private -> public 변경한 메서드 테스트 코드 추가

cutehumanS2 committed:
feat: 재입력 로직 구현, 책임 분리

-게임 결과 산출 책임을 GameStatus에게 위임
-게임 턴 관리 책임을 ChessGame에게 위임

jinyoungchoi95 commented on src/main/java/chess/controller/ChessGameController.java:
```diff
@@ -0,0 +1,125 @@
+import java.util.Optional;
+import java.util.function.Supplier;
+
+public class ChessGameController {
```
클래스 이름을 Controller로 바꿔주신 것 같은데요 Controller라는 이름이 적절할까요?

MVC, Spring에서 사용하는 controller를 생각하고 네이밍하셨다면 1레벨은 "자바만을 사용하여 만드는 어플리케이션"이기 때문에 특정 패턴에 종속적인 네이밍을 사용하지 않아도 됩니다. 패턴은 이론적으로 학습하고 적용할 것을 찾아 억지로 네이밍하는 것보다는 자연스럽게 미션을 통해서 적절한 상황의 디자인 패턴을 발견하는 재미를 느끼셨으면 좋겠습니다 :)

그럼 종속적인 네이밍을 뒤로하고 "컨트롤러"라는 이름은 이 클래스의 기능을 담아내기에 충분한 네이밍을 가지고 있나요? 컨트롤러는 조이스틱과 같은 게임의 컨트롤러를 뜻한다고 보이는데요. 조이스틱이 게임을 실행시키고 화면을 출력하지 않듯 Controller보다는 이 객체를 표현해줄 더 좋은 네이밍을 고려하셨으면 좋겠습니다.

cutehumanS2 replied:
‘Controller’라는 이름의 적절성에 대해 생각해 본 적이 없는데, 
오리의 의견을 읽고 나니 무분별하게 이름을 지어서는 안 되겠다는 생각이 듭니다. 
확실히 ‘Controller’라는 이름은 적절하지 않은 것 같아요.

그래서 아래와 같이 객체들의 이름을 변경해 보았습니다. 적절한 네이밍… 쉽지 않네요.🥺
> ChessGameController → ChessGame로 재변경
> ChessGame → GameStatus
> GameStatus → GameResult

jinyoungchoi95 commented on src/main/java/chess/domain/square/Rank.java:
```diff
@@ -56,20 +56,16 @@ public List<Rank> findRankPath(final Rank other) {
     }
 
-    private Rank valueOfIndex(final int value) {
+    public static Rank findByIndex(final int index) {
```
냥인은 static과 일반 메소드의 위치를 어떻게 배치하시는 편인가요?

cutehumanS2 replied:
보통은 static 메소드를 먼저 배치한 후 일반 메소드를 배치합니다.

`findByIndex()` 의 경우 원래 Rank 클래스 안에서만 쓰이던 private 메소드였는데, 
public static으로 변경하면서 배치도 바꿔 주는 걸 놓쳤습니다.😅

메서드 순서 변경하였습니다!

jinyoungchoi95 commented on src/main/java/chess/view/Command.java:
```diff
@@ -32,20 +33,4 @@ public static void validateFormat(final String value) {
         }
     }
-
-    public boolean isStart() {
```
제거하신 이유가 있을까요?

cutehumanS2 replied:
Command 종류마다 `isXXX()`와 같이 메서드를 만들지 않고도, 외부에서 `==` 비교를 해주어도 괜찮다고 생각해서 지우게 되었습니다(왜 괜찮다고 생각했는지 근거는 사실 없습니다😹). 

> enum인데 index비교를 하지 않고 `==`을 외부에서 호출해도되지 않나요?

1,2단계 때 위와 같은 리뷰를 받기도 했고요!

혹시 `isXXX()` 와 같이 메서드를 만들어 사용하는 것을 권장하신다면 이유를 알 수 있을까요?

jinyoungchoi95 replied:
권장한다기보단 삭제하셨길래 삭제한 이유에 대해 여쭈어보았던 것이었어요. 이전에 말씀드린대로 외부에서 `==` 비교를 해도 충분하다고 생각합니다 :)

jinyoungchoi95 commented on src/main/java/chess/domain/piece/strategy/MoveStrategy.java:
```diff
@@ -1,8 +1,9 @@
 public interface MoveStrategy {
 
-    boolean canMove(final Square source, final Square target);
+    boolean canMove(final Board board, final Square source, final Square target);
```
👍

jinyoungchoi95 commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -45,8 +45,8 @@ private void validateIsNonExistentPiece(final Square square) {
 
-    private Piece findPieceBySquare(final Square square) {
-        return pieces.get(square);
+    public Piece findPieceBySquare(final Square square) {
```
public으로 변경되면서 메소드의 위치 조정이 필요해보여요.

jinyoungchoi95 commented on src/main/java/chess/domain/game/ChessGame.java:
```diff
@@ -0,0 +1,30 @@
+        this.turn = turn;
+    }
+
+    public void move(final Square source, final Square target) {
+        board.move(source, target, turn);
+        this.turn = turn.next();
+    }
```
turn이 변경되었는지에 대한 테스트가 필요해보여요.

cutehumanS2 replied:
턴이 변경되었는지에 대한 테스트를 아래와 같이 작성했습니다.

ChessGame에 turn의 getter가 없어서, turn에 저장된 값을 직접 검증하는 대신
‘어떤 진영이 연속으로 기물을 이동하려는 경우 예외가 발생하는지’를 검증하는 식으로 작성하였습니다. 오리는 이러한 방식으로 테스트를 작성하는 것에 대해 어떻게 생각하시나요? 괜찮은 방법인 것 같으신가요?
```java
  @DisplayName("기물을 이동하면 턴이 바뀐다.")
    @Test
    void changeTurnWhenMovePiece() {
        final ChessGame chessGame = new ChessGame(PieceColor.WHITE);
        chessGame.move(new Square(File.e, Rank.TWO), new Square(File.e, Rank.FOUR));

        assertThatThrownBy(() -> chessGame.move(new Square(File.e, Rank.FOUR), new Square(File.e, Rank.FIVE)))
                .isInstanceOf(IllegalArgumentException.class);
    }
```

jinyoungchoi95 replied:
현재 구조상 이정도로도 충분한 것 같아요. 더 단순하고 로직이 안타는 테스트를 작성한다면 `ChessStatus`에 equals를 재정의하여 테스트해보는 방법도 있겠네요.

다만 앞으로도 고민해보시면 좋을 내용은 어떻게 단순하게 테스트할 수 있느냐에요. 단순하게 테스트할 수 있다는건 마냥 테스트를 단순하게 짜라는 것이 아닌 테스트를 단순하게 짤수있는 구조로 설계하라는 것을 뜻해요. `ChessStatus`라는 단위테스트가 테스트하기 불편한 구조라면 어떻게 구조 개선을 해야할지도 혹은 현재 클래스에서 개선이 안보인다면 클래스자체의 문제는 아닐지 앞으로 다른 미션을 하실때도 고민해보시면 도움이 될 것 같아요. (현재가 개선이 안보인다거나 구조개선이 필요하다는 이야기는 아니에요.)

cutehumanS2 replied:
‘테스트하기 좋은 설계’, ‘좋은 구조가 테스트 하기 쉽다.’ 
앞으로 꼭 염두에 두며 설계해야겠군요. 감사합니다. 🙌

jinyoungchoi95 commented on src/main/java/chess/domain/game/GameStatus.java:
```diff
@@ -0,0 +1,101 @@
+        return pieces.values().stream()
+                .filter(piece -> !piece.isSameType(PieceType.PAWN))
+                .filter(piece -> piece.isSameColor(color))
+                .mapToDouble(piece -> piece.getType().getScore())
```
```suggestion
                .mapToDouble(Piece::getScore)
```
만들어주신 메소드로 대체가 가능해보이네요.

cutehumanS2 replied:
그렇네요. 말씀해 주신 대로 변경하였습니다. 😸

jinyoungchoi95 commented on src/main/java/chess/domain/game/GameStatus.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    private Map<File, List<Piece>> findPawnsByFile(final PieceColor pawnColor) {
+        return pieces.entrySet().stream()
+                .filter(piece -> piece.getValue().isSameType(PieceType.PAWN))
+                .filter(piece -> piece.getValue().isSameColor(pawnColor))
```
`entirySet`에서 stream으로 내려오는값은 `Entry<Square, Piece>`이기 때문에 각 필터에서 이 값들을 `piece`라고는 보기 힘들지 않을까요?

cutehumanS2 replied:
오리의 의견이 맞습니다. 
`piece → entry`로 변경하였습니다!

jinyoungchoi95 commented on src/main/java/chess/domain/game/GameStatus.java:
```diff
@@ -0,0 +1,101 @@
+        return pawnBasicScore * count;
+    }
+
+    private Piece getPawn(List<Piece> pawns) {
+        return pawns.get(0);
+    }
```
어떤 의도로 만드셨고, 의미상도 사실 맞는말이긴하지만 `List<Piece>.get(0)`을 했다라는 메소드의 내용을 다른사람이 보았을 때 쉽게 이해가가능할지 생각해보시면 좋을 것 같아요.

이름만 pawns이지 `List<Piece>`에서 index 0을 가져오는게 pawn을 가져오는 행위가 맞을까요? 다른 `Piece`구현체가 들어있으면 이 메소드의 결과가 어떻게 되나요?

cutehumanS2 replied:
해당 코드가 이해하기 좋은 코드가 아닌 점, 인정합니다. 
저도 이 부분에 대해 굉장히 찜찜했거든요. ㅎ ㅎ 

List<Piece>로 pawn이 아닌 다른 Piece 구현체가 들어오면 결과가 의도한 바와 달라질 수 있다는 점은 미처 고려하지 못했습니다.

``List<Piece>.get(0)``과 같은 코드는 필요하지 않도록 리팩토링 하였습니다.

jinyoungchoi95 commented on src/main/java/chess/domain/game/GameStatus.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    private PieceColor findFinalWinnerTeam() {
+        return pieces.values().stream()
```
> 한 줄에 점을 하나만 찍는다.

해당 규칙이 지켜지지 않은 코드들이 다른곳에도 보여요.

cutehumanS2 replied:
저번부터 리뷰 받았던 내용인데 여전히 잘 안 지켜지고 있었네요. 죄송합니다.😿

한 줄에 점은 하나만 찍도록 최대한 노력해 봤는데, 아래와 같이 지키지 않은 부분도 아직 존재합니다(이 부분들은 개선할 수 없거나 그냥 두는 편이 더 낫다고 생각했습니다)!
```java
// File.findByValue()
    public static File findByValue(final String value) {
        return Arrays.stream(values())
                .filter(file -> file.name().equals(value))
                .findAny()
                .orElseThrow(() -> new IllegalArgumentException(value + ERROR_NOT_EXIST_FILE));
    }
```
```java
// Rank.validateNumeric()
    private static void validateNumeric(final String rank) {
        if (!NUMBER_PATTERN.matcher(rank).matches()) {
            throw new IllegalArgumentException(ERROR_NOT_NUMERIC);
        }
    }
```
```java
// Command.isMatchedCommandFormat()
    private boolean isMatchedCommandFormat(final Pattern format, final String value) {
        return format.matcher(value).matches();
    }
```

jinyoungchoi95 commented on src/main/java/chess/controller/ChessGameController.java:
```diff
@@ -0,0 +1,125 @@
+        if (command == Command.MOVE) {
+            move(game, commandInput);
+
+            if (gameStatus.isGameOver()) {
+                return Command.END;
+            }
```
> indent(인덴트, 들여쓰기) depth를 2를 넘지 않도록 구현한다. 1까지만 허용한다.

cutehumanS2 replied:
해당 조건문을  같은 클래스 내에 있는 `move()` 메소드로 옮겨 인덴트를 줄여보았습니다.😸

jinyoungchoi95 reviewed (changes requested):
안녕하세요 냥인 😄

3단계 잘 구현해주셨네요. 몇가지 코멘트 남겨두었으니 확인부탁드려요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

cutehumanS2 committed:
refactor: 메서드 배치 변경

cutehumanS2 committed:
refactor: 메서드명 변경, 불필요한 코드 및 공백 제거

cutehumanS2 committed:
refactor: 점수 계산 로직 변경

-폰 점수에 적용되는 패널티의 상수명 변경

cutehumanS2 committed:
refactor: 게임 종료 조건 위치 변경, 출력문 변경

cutehumanS2 committed:
refactor: 메서드 분리

cutehumanS2 committed:
test: 턴 변경 여부 테스트

cutehumanS2 committed:
refactor: 네이밍 변경

-ChessGameController → ChessGame
-ChessGame → GameStatus
-GameStatus → GameResult

jinyoungchoi95 commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -95,6 +78,11 @@ private void replacePieceSquare(final Square source, final Square target) {
+    public Piece findPieceBySquare(final Square square) {
+        return pieces.getOrDefault(square, new Piece(PieceType.EMPTY, PieceColor.NONE));
+    }
+
+
```
```suggestion

```
👀

jinyoungchoi95 reviewed (approved):
안녕하세요 냥인 😄

피드백 잘 반영해주셨네요. 다음단계 진행을 위해 머지하겠습니다.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️
